git_theta/utils.py

```diff
@@ -34,3 +45,118 @@ def _iterate_dict_leaves(d, prefix):
                 yield (v, prefix + [k])
 
     return _iterate_dict_leaves(d, [])
+
+
+def reconstruct_apply_update(updates_folder, update_file, update_type, base_value):
+    """
+    Reconstruct dense update from other update types and apply that to base_value
+    Parameters
+    ----------
+    updates_folder
+    update_file
+    update_type
+    base_value
+
+    Returns
+    -------
+    updated_val
+    """
+    updated_val = base_value
+    if update_type == constants.SPARSE_UPDATE:
+        # Reshape since shape sometimes changes on load. e.g, actual shape (128,) on load becomes (1,128)
+        update_val = (
+            sparse.load_npz(os.path.join(updates_folder, update_file))
+            .toarray()
+            .reshape(base_value.shape)
+        )
+        np.copyto(updated_val, update_val, where=update_val != 0)
+
+    return updated_val
+
+
+def apply_param_updates(base_val, updates_folder):
+    """
+    Iterate through all update files and apply them to the dense base_val
+    Parameters
+    ----------
+    base_val
+    updates_folder
+
+    Returns
+    -------
+    final_val
+    """
+    if not os.path.exists(updates_folder) or not os.listdir(updates_folder):
+        logging.debug(f"No updates in {updates_folder}")
+        return base_val
+
+    final_val = base_val
+    with open(
+        os.path.join(updates_folder, constants.THETA_UPDATES_METADATA_FILE), "r"
+    ) as f:
+        update_apply_order = [line.strip("\n").split(",") for line in f.readlines()]
+
+    logging.info(f"{len(update_apply_order)} updates found in {updates_folder}")
+    for update_file, update_type in update_apply_order:
+        logging.info(f"Applying update from {update_file}")
+        final_val = reconstruct_apply_update(
+            updates_folder, update_file, update_type, final_val
+        )
+
+    return final_val
+
+
+def get_prev_param_dict(theta_model_dir):
```

Comment:
replace the for loop in the smudge filter with this function to "checkout" latest committed state of model